### ISSUE TITLE
How to Develop a Skillful Machine Learning Time Series Forecasting Model

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#438]
How to Develop a Skillful Machine Learning Time Series Forecasting Model<br>
<br>
https://ift.tt/2nmyoY6<br>
<br>
Jason Brownlee<br>
<br>
<p>You are handed data and told to develop a forecast model.</p>
<p><em>What do you do?</em></p>
<p>This is a common situation; far more common than most people think.</p>
<ul>
<li>Perhaps you are sent a CSV file.</li>
<li>Perhaps you are given access to a database.</li>
<li>Perhaps you are starting a competition.</li>
</ul>
<p>The problem can be reasonably well defined:</p>
<ul>
<li>You have or can access historical time series data.</li>
<li>You know or can find out what needs to be forecasted.</li>
<li>You know or can find out how what is most important in evaluating a candidate model.</li>
</ul>
<p>So how do you tackle this problem?</p>
<p>Unless you have been through this trial by fire, you may struggle.</p>
<ul>
<li>You may struggle because you are new to the fields of machine learning and time series.</li>
<li>You may struggle even if you have machine learning experience because time series data is different.</li>
<li>You may struggle even if you have a background in time series forecasting because machine learning methods may outperform the classical approaches on your data.</li>
</ul>
<p>In all of these cases, you will benefit from working through the problem carefully and systematically.</p>
<p>In this post, I want to give you a specific and actionable procedure that you can use to work through your time series forecasting problem.</p>
<p>Let&rsquo;s get started.</p>
<div class="wp-caption">
<img src="https://machinelearningmastery.com/wp-content/uploads/2018/08/How-to-Develop-a-Skilful-Time-Series-Forecasting-Model.jpg" alt="How to Develop a Skilful Time Series Forecasting Model" width="640" class="wp-image-5982" height="277">
<p class="wp-caption-text">How to Develop a Skilful Time Series Forecasting Model<br>
Photo by <a href="https://www.flickr.com/photos/makeitkenya/22036888582/">Make it Kenya</a>, some rights reserved.</p>
</div>
<h2>Process Overview</h2>
<p>The goal of this process is to get a &ldquo;<em>good enough</em>&rdquo; forecast model as fast as possible.</p>
<p>This process may or may not deliver the best possible model, but it will deliver a good model: a model that is better than a baseline prediction, if such a model exists.</p>
<p>Typically, this process will deliver a model that is 80% to 90% of what can be achieved on the problem.</p>
<p>The process is fast. As such, it focuses on automation. Hyperparameters are searched rather than specified based on careful analysis. You are encouraged to test suites of models in parallel, rapidly getting an idea of what works and what doesn&rsquo;t.</p>
<p>Nevertheless, the process is flexible, allowing you to circle back or go as deep as you like on a given step if you have the time and resources.</p>
<p>This process is divided into four parts; they are:</p>
<ol>
<li>Define Problem</li>
<li>Design Test Harness</li>
<li>Test Models</li>
<li>Finalize Model</li>
</ol>
<p>You will notice that the process is different from a classical linear work-through of a predictive modeling problem. This is because it is designed to get a working forecast model fast and then slow down and see if you can get a better model.</p>
<p>What is your process for working through a new time series forecasting problem?<br>
Share it below in the comments.</p>
<h2>How to Use This Process</h2>
<p>The biggest mistake is skipping steps.</p>
<p>For example, the mistake that almost all beginners make is going straight to modeling without a strong idea of what problem is being solved or how to robustly evaluate candidate solutions. This almost always results in a lot of wasted time.</p>
<p>Slow down, follow the process, and complete each step.</p>
<p>I recommend having separate code for each experiment that can be re-run at any time.</p>
<p>This is important so that you can circle back when you discover a bug, fix the code, and re-run an experiment. You are running experiments and iterating quickly, but if you are sloppy, then you cannot trust any of your results. This is especially important when it comes to the design of your test harness for evaluating candidate models.</p>
<p>Let&rsquo;s take a closer look at each step of the process.</p>
<h2>1. Define Problem</h2>
<p>Define your time series problem.</p>
<p>Some topics to consider and motivating questions within each topic are as follows:</p>
<ol>
<li>Inputs vs. Outputs
<ol>
<li>What are the inputs and outputs for a forecast?</li>
</ol>
</li>
<li>Endogenous vs. Exogenous
<ol>
<li>What are the endogenous and exogenous variables?</li>
</ol>
</li>
<li>Unstructured vs. Structured
<ol>
<li>Are the time series variables unstructured or structured?</li>
</ol>
</li>
<li>Regression vs. Classification
<ol>
<li>Are you working on a regression or classification predictive modeling problem?</li>
<li>What are some alternate ways to frame your time series forecasting problem?</li>
</ol>
</li>
<li>Univariate vs. Multivariate
<ol>
<li>Are you working on a univariate or multivariate time series problem?</li>
</ol>
</li>
<li>Single-step vs. Multi-step
<ol>
<li>Do you require a single-step or a multi-step forecast?</li>
</ol>
</li>
<li>Static vs. Dynamic
<ol>
<li>Do you require a static or a dynamically updated model?</li>
</ol>
</li>
</ol>
<p>Answer each question even if you have to estimate or guess.</p>
<p>Some useful tools to help get answers include:</p>
<ul>
<li>Data visualizations (e.g. line plots, etc.).</li>
<li>Statistical analysis (e.g. ACF/PACF plots, etc.).</li>
<li>Domain experts.</li>
<li>Project stakeholders.</li>
</ul>
<p>Update your answers to these questions as you learn more.</p>
<h2>2. Design Test Harness</h2>
<p>Design a test harness that you can use to evaluate candidate models.</p>
<p>This includes both the method used to estimate model skill and the metric used to evaluate predictions.</p>
<p>Below is a common time series forecasting model evaluation scheme if you are looking for ideas:</p>
<ol>
<li>Split the dataset into a train and test set.</li>
<li>Fit a candidate approach on the training dataset.</li>
<li>Make predictions on the test set directly or using walk-forward validation.</li>
<li>Calculate a metric that compares the predictions to the expected values.</li>
</ol>
<p>The test harness must be robust and you must have complete trust in the results it provides.</p>
<p>An important consideration is to ensure that any coefficients used for data preparation are estimated from the training dataset only and then applied on the test set. This might include mean and standard deviation in the case of data standardization.</p>
<h2>3. Test Models</h2>
<p>Test many models using your test harness.</p>
<p>I recommend carefully designing experiments to test a suite of configurations for standard models and letting them run. Each experiment can record results to a file, to allow you to quickly discover the top three to five most skilful configurations from each run.</p>
<p>Some common classes of methods that you can design experiments around include the following:</p>
<ul>
<li>Baseline.
<ul>
<li>Persistence (grid search the lag observation that is persisted)</li>
<li>Rolling moving average.</li>
<li>&hellip;</li>
</ul>
</li>
<li>Autoregression.
<ul>
<li>ARMA for stationary data.</li>
<li>ARIMA for data with a trend.</li>
<li>SARIMA for data with seasonality.</li>
<li>&hellip;</li>
</ul>
</li>
<li>Exponential Smoothing.
<ul>
<li>Simple Smoothing</li>
<li>Holt Winters Smoothing</li>
<li>&hellip;</li>
</ul>
</li>
<li>Linear Machine Learning.
<ul>
<li>Linear Regression</li>
<li>Ridge Regression</li>
<li>Lasso Regression</li>
<li>Elastic Net Regression</li>
<li>&hellip;.</li>
</ul>
</li>
<li>Nonlinear Machine Learning.
<ul>
<li>k-Nearest Neighbors</li>
<li>Classification and Regression Trees</li>
<li>Support Vector Regression</li>
<li>&hellip;</li>
</ul>
</li>
<li>Ensemble Machine Learning.
<ul>
<li>Bagging</li>
<li>Boosting</li>
<li>Random Forest</li>
<li>Gradient Boosting</li>
<li>&hellip;</li>
</ul>
</li>
<li>Deep Learning.
<ul>
<li>MLP</li>
<li>CNN</li>
<li>LSTM</li>
<li>Hybrids</li>
<li>&hellip;</li>
</ul>
</li>
</ul>
<p>This list is based on a univariate time series forecasting problem, but you can adapt it for the specifics of your problem, e.g. use VAR/VARMA/etc. in the case of multivariate time series forecasting.</p>
<p>Slot in more of your favorite classical time series forecasting methods and machine learning methods as you see fit.</p>
<p>Order here is important and is structured in increasing complexity from classical to modern methods. Early approaches are simple and give good results fast; later approaches are slower and more complex, but also have a higher bar to clear to be skillful.</p>
<p>The resulting model skill can be used in a ratchet. For example, the skill of the best persistence configuration provide a baseline skill that all other models must outperform. If an autoregression model does better than persistence, it becomes the new level to outperform in order for a method to be considered skilful.</p>
<p>Ideally, you want to exhaust each level before moving on to the next. E.g. get the most out of Autoregression methods and use the results as a new baseline to define &ldquo;skilful&rdquo; before moving on to Exponential Smoothing methods.</p>
<p>I put deep learning at the end as generally neural networks are poor at time series forecasting, but there is still a lot of room for improvement and experimentation in this area.</p>
<p>The more time and resources that you have, the more configurations that you can evaluate.</p>
<p>For example, with more time and resources, you could:</p>
<ul>
<li>Search model configurations at a finer resolution around a configuration known to already perform well.</li>
<li>Search more model hyperparameter configurations.</li>
<li>Use analysis to set better bounds on model hyperparameters to be searched.</li>
<li>Use domain knowledge to better prepare data or engineer input features.</li>
<li>Explore different potentially more complex methods.</li>
<li>Explore ensembles of well performing base models.</li>
</ul>
<p>I also encourage you to include data preparation schemes as hyperparameters for model runs.</p>
<p>Some methods will perform some basic data preparation, such as differencing in ARIMA, nevertheless, it is often unclear exactly what data preparation schemes or combinations of schemes are required to best present a dataset to a modeling algorithm. Rather than guess, grid search and decide based on real results.</p>
<p>Some data preparation schemes to consider include:</p>
<ul>
<li>Differencing to remove a trend.</li>
<li>Seasonal differencing to remove seasonality.</li>
<li>Standardize to center.</li>
<li>Normalize to rescale.</li>
<li>Power Transform to make normal.</li>
</ul>
<p>So much searching can be slow.</p>
<p>Some ideas to speed up the evaluation of models include:</p>
<ul>
<li>Use multiple machines in parallel via cloud hardware (such as Amazon EC2).</li>
<li>Reduce the size of the train or test dataset to make the evaluation process faster.</li>
<li>Use a more coarse grid of hyperparameters and circle back if you have time later.</li>
<li>Perhaps do not refit a model for each step in walk-forward validation.</li>
</ul>
<h2>4. Finalize Model</h2>
<p>At the end of the previous time step, you know whether your time series is predictable.</p>
<p>If it is predictable, you will have a list of the top 5 to 10 candidate models that are skillful on the problem.</p>
<p>You can pick one or multiple models and finalize them. This involves training a new final model on all available historical data (train and test).</p>
<p>The model is ready for use; for example:</p>
<ul>
<li>Make a prediction for the future.</li>
<li>Save the model to file for later use in making predictions.</li>
<li>Incorporate the model into software for making predictions.</li>
</ul>
<p>If you have time, you can always circle back to the previous step and see if you can further improve upon the final model.</p>
<p>This may be required periodically if the data changes significantly over time.</p>
<h2>Further Reading</h2>
<p>This section provides more resources on the topic if you are looking to go deeper.</p>
<ul>
<li><a href="https://machinelearningmastery.com/time-series-forecasting/">What Is Time Series Forecasting?</a></li>
<li><a href="https://machinelearningmastery.com/work-time-series-forecast-project/">How to Work Through a Time Series Forecast Project</a></li>
<li><a href="https://machinelearningmastery.com/train-final-machine-learning-model/">How to Train a Final Machine Learning Model</a></li>
</ul>
<h2>Summary</h2>
<p>In this post, you discovered a simple four-step process that you can use to quickly discover a skilful predictive model for your time series forecasting problem.</p>
<p>Did you find this process useful?<br>
Let me know below.</p>
<p>Do you have any questions?<br>
Ask your questions in the comments below and I will do my best to answer.</p>
<p>The post <a rel="nofollow" href="https://machinelearningmastery.com/how-to-develop-a-skilful-time-series-forecasting-model/">How to Develop a Skillful Machine Learning Time Series Forecasting Model</a> appeared first on <a rel="nofollow" href="https://machinelearningmastery.com">Machine Learning Mastery</a>.</p>
<br>
<br>
<br>
<br>
AI|ML|DM|IR|PR<br>
<br>
via Machine Learning Mastery https://ift.tt/2if3vo8<br>
<br>
August 10, 2018 at 03:03AM
